Yank/restraints.py

```diff
@@ -1992,6 +2048,285 @@ def _assign_if_undefined(attr_name, attr_value):
+        or :func:`Topography Select String <yank.Topography.select>`.
+        Any number of receptor atoms can be selected.
+        This can temporarily be left undefined, but ``determine_missing_parameters()``
+        must be called before using the Restraint object. The same if a DSL
+        expression or Topography region is provided.
+        If no selection is given, all receptor atoms will be restrained.
+        If an empty list is provided, no receptor atoms will be restrained.
+        (default is None).
+    restrained_ligand_atoms : iterable of int, str, or None; Optional
+        The indices of the ligand atoms to restrain, an MDTraj DSL expression, or a
+        :class:`Topography <yank.Topography>` region name,
+        or :func:`Topography Select String <yank.Topography.select>`.
+        Any number of ligand atoms can be selected
+        This can temporarily be left undefined, but ``determine_missing_parameters()``
+        must be called before using the Restraint object. The same if a DSL
+        expression or Topography region is provided
+        If no selection is given, all ligand atoms will be restrained.
+        If an empty list is provided, no receptor atoms will be restrained.
+        (default is None).
+    K_RMSD : simtk.unit.Quantity, optional, default=0.6*kilocalories_per_mole/angstrom**2
+        The spring constant (units compatible with kilocalories_per_mole/angstrom**2).
+    RMSD0 : simtk.unit.Quantity, optional, default=2.0*angstrom
+        The RMSD at which the restraint becomes nonzero.
+
+    Attributes
+    ----------
+    restrained_receptor_atoms : list of int
+        The indices of the restrained receptor atoms
+    restrained_ligand_atoms : list of int
+        The indices of the restrained_ligand_atoms
+
+    Examples
+    --------
+    Create the ThermodynamicState.
+
+    >>> from openmmtools import testsystems, states
+    >>> system_container = testsystems.LysozymeImplicit()
+    >>> system, positions = system_container.system, system_container.positions
+    >>> thermodynamic_state = states.ThermodynamicState(system, 298*unit.kelvin)
+    >>> sampler_state = states.SamplerState(positions)
+
+    Identify ligand atoms. Topography automatically identify receptor atoms too.
+
+    >>> from yank.yank import Topography
+    >>> topography = Topography(system_container.topology, ligand_atoms=range(2603, 2621))
+
+    Create a restraint
+
+    >>> restraint = RMSD(restrained_receptor_atoms=[1335, 1339, 1397],
+    ...                  restrained_ligand_atoms=[2609, 2607, 2606],
+    ...                  K_RMSD=1.0*unit.kilocalories_per_mole/unit.angstrom**2,
+    ...                  RMSD0=1*unit.angstroms)
+
+    Get standard state correction.
+
+    >>> correction = restraint.get_standard_state_correction(thermodynamic_state)
+
+    """
+    def __init__(self, restrained_receptor_atoms=None, restrained_ligand_atoms=None,
+                 K_RMSD=None, RMSD0=None):
```

Comment:
Can you maybe add a `reference_positions=None` argument here? It makes sense if people use the API to be able to instantiate a restraint without instantiating a `SamplerState` and a `Topography` object first.

For curiosity, do you know why OpenMM requires passing the positions of all the particles to `RMSDForce` instead of only the restrained ones?

Reply:
I thought about making it public, but wanted to ask about it. I can make it public

It needs a full reference coordinate set to know how the atoms are placed in the world relative to the system itself. Its much much easier to engineer a solution where you pass the same number of particles to the system as every other force and uses the same atomic indices (e.g. atom 512 is Force A is the same as 512 in Force B)